Rebase SILScope generation on top of ASTScope.

adrian-prantl opened the pull request:
Rebase SILScope generation on top of ASTScope.

This patch replaces the stateful generation of SILScope information in SILGenFunction with data derived from the ASTScope hierarchy, which should be 100% in sync with the scopes needed for local variables. The goal is to eliminate the surprising effects that the stack of cleanup operations can have on the current state of SILBuilder leading to a fully deterministic (in the sense of: predictible by a human) association of SILDebugScopes with SILInstructions. The patch also eliminates the need to many workarounds. Thereare still some accomodations for several Sema transformation passes such as ResultBuilders, which don't correctly update the source locations when moving round nodes. If these were implemented as macros, this problem would disappear.

This necessary rewrite of the macro scope handling included in this patch also adds proper support nested macro expansions.

This fixes

rdar://88274783
and either fixes or at least partially addresses the following:

rdar://89252827
rdar://105186946
rdar://105757810
rdar://105997826
rdar://105102288

___

Cherry pick of #64551.

adrian-prantl committed:
SILLocation: Move pointsToEnd flag into ASTNodeTy. NFC

The pointsToEnd flag is only meaningful when combined with an AST node, and
moving it into the AST pointer frees up a bit in the flags bitfield for later
use.

(cherry picked from commit 63304d078766b20a51c2368b246fd38a903d305f)

adrian-prantl committed:
Preserve the Implicit attribute in SILLocation. (NFC)

This change aims at reducing the need for SIL passes to check into the AST
storage of SILLocations. The end goal is to eventually merge this with the
autogenerated flag, but at the moment the emergent semantics of both properties
are not identical.

(cherry picked from commit c5efebe7840f043beae10dd2c29f431d604faf73)

adrian-prantl committed:
Rebase SILScope generation on top of ASTScope.

This patch replaces the stateful generation of SILScope information in
SILGenFunction with data derived from the ASTScope hierarchy, which should be
100% in sync with the scopes needed for local variables. The goal is to
eliminate the surprising effects that the stack of cleanup operations can have
on the current state of SILBuilder leading to a fully deterministic (in the
sense of: predictible by a human) association of SILDebugScopes with
SILInstructions. The patch also eliminates the need to many workarounds. There
are still some accomodations for several Sema transformation passes such as
ResultBuilders, which don't correctly update the source locations when moving
around nodes. If these were implemented as macros, this problem would disappear.

This necessary rewrite of the macro scope handling included in this patch also
adds proper support nested macro expansions.

This fixes

rdar://88274783

and either fixes or at least partially addresses the following:

rdar://89252827
rdar://105186946
rdar://105757810
rdar://105997826
rdar://105102288
(cherry picked from commit 158772c2ab0fcc9cd9882c4c46e2e592ae306420)

adrian-prantl committed:
Remove uses of SGF.CurrentSILLoc from SILGenPattern

(cherry picked from commit 0337ea94f207ef3971067a361be3c794e3256de2)

adrian-prantl committed:
Refactor testcase to be more robust against IR ordering changes

and also make it easier to debug.

(cherry picked from commit bbcce343e81c6c01f057ac48992f247c6a478ee7)

adrian-prantl commented:
@swift-ci test

adrian-prantl committed:
Add a test for nested macro debug info

rdar://105997826
(cherry picked from commit 475f7798cd28576cd7e70fab2b00e8e1f3e09111)

adrian-prantl commented:
test with https://github.com/apple/llvm-project/pull/6611
@swift-ci test

DougGregor commented on lib/SILGen/SILGenEpilog.cpp:
```diff
@@ -319,6 +319,11 @@ emitEpilog(SILLocation TopLevel, bool UsesCustomEpilog) {
     // Return () if no return value was given.
     if (!returnValue)
       returnValue = emitEmptyTuple(CleanupLocation(TopLevel));
+    //{
+    //  RegularLocation loc(returnLoc);
+    //  loc.pointToEnd();
+    //  returnValue = emitEmptyTuple(loc);
+    //}
```
Should this be uncommented or removed?

adrian-prantl replied:
Thanks for noticing!

adrian-prantl committed:
Delete commented-out code. NFC

(cherry picked from commit e3445a8e733518ccbf7d4ab88e2a44ffda123d60)

adrian-prantl committed:
Ignore profile counter instructions in the dihole verifier.

rdar://107764966
(cherry picked from commit c877a4a802ec13c5463ddd235853c0e3d21f82e2)

adrian-prantl committed:
Fix handling of implicit locations for variables

(cherry picked from commit bccc080888f47f414da22d917f486566d2db10c9)

adrian-prantl committed:
Don't emit freestanding macros in inlined scopes

(cherry picked from commit e51937f29e27f213a6de489367790e16838f569d)

adrian-prantl committed:
Skip instructions without source location in dihole verfication

(cherry picked from commit 8d6c55e673e14183e0179a346d1e5fa2a106d401)

adrian-prantl commented:
@swift-ci test

adrian-prantl commented:
@swift-ci test

adrian-prantl commented:
@swift-ci test

adrian-prantl committed:
Update the scope of SILBuilder in ClosureCloner.

This is something that every visit() function needs to do and these manual
overloads forgot to. I tried to come up with a more general solution, but due
the rather nested inheritance of SILCloner this turned out harde than expected.

rdar://107984038
(cherry picked from commit fbf4c92bbac81736d7508c6117cb99a4f1a5a189)

adrian-prantl commented:
@swift-ci test

adrian-prantl commented:
@swift-ci test

adrian-prantl committed:
SILgenPattern: Mark location used by forwardIntoSubtree as autogenerated.

If a pattern reuses a value from a previous branch this could result in a step
backwords or a previous scope being reopeneed.

rdar://107764966
(cherry picked from commit 68aa944ccd02324cfbb595a4d822e4cbf68131b1)

adrian-prantl committed:
Disable dihole verifier in a non-asserts build

(cherry picked from commit b6be6ad809bba1fcca201f04c3201f60e702023e)

adrian-prantl commented:
Fix one more bug that was uncovered by the di-hole verification when building the source compatibility suite in debug mode.

adrian-prantl commented:
@swift-ci test

adrian-prantl commented:
@swift-ci test source compatibility

adrian-prantl commented:
Source compatibility failures are unrelated to this PR.
